File: Tercera Entrega/sdi3-37.Cli-REST/src/com/sdi/presentation/user/action/MarcarTareaFinalizadaAction.java
package com.sdi.presentation.user.action;

import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.client.Entity;
import javax.ws.rs.core.GenericType;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import com.sdi.dto.Task;
import com.sdi.presentation.user.Sesion;

import alb.util.console.Console;
import alb.util.menu.Action;

public class MarcarTareaFinalizadaAction implements Action {

	public MarcarTareaFinalizadaAction() {}
	
	@Override
	public void execute() throws Exception {
		Console.println("------MARCAR TAREA COMO FINALIZADA------");
		
		//Pedir datos
		String tareaId = Console.readString("Id de la tarea que desea finalizar");
		
		//Comprobamos que nos hayan metido un número y no letras o caracteres
		if(isNumeric(tareaId)) {
			//Buscamos la tarea por el id
			GenericType<Task> modelo = new GenericType<Task>() {};
			
			Task task = ClientBuilder.newClient()				
					.target(Sesion.getInstance().getRestServiceUrl())
					.path("buscarTarea")				
					.path("id="+tareaId)
					.request()
					.get()
					.readEntity(modelo);
			
			//Comprobamos que exista la tarea
			if(task != null && (long) task.getUserId() == Sesion.getInstance().getUser().getId()) {
				//Comprobamos que no este finalizada
				if(task.getFinished() == null) {
					Response response = ClientBuilder.newClient()				
							.target(Sesion.getInstance().getRestServiceUrl())
							.path("finalizarTarea")	
							.request()
							.put(Entity.entity(new Task(task.getId()), MediaType.APPLICATION_JSON));
								
					if(response.getStatus() == 204) {
						Console.println("\tTarea con id "+tareaId+" marcada como finalizada correctamente");
					} 
					else {
						Console.println("\tHubo algún problema al marcar como finalizada la tarea con id "+tareaId);
					}
				}
				else {
					Console.println("\tLa tarea con id "+tareaId+ " ya estaba finalizada");
				}
			}
			else {				
				Console.println("\tNo existe la tarea con id "+tareaId);
			}
		}
		else {
			Console.println("\tId introducido inválido");
		}
		
		Console.println("----------------------------------------");
	}
	
	private boolean isNumeric(String cadena) {
		try {
			Integer.parseInt(cadena);
			return true;
		}
		catch (NumberFormatException e) {
			return false;
		}
	}
}

File: Tercera Entrega/sdi3-37.Cli-REST-C_Sharp/sdi3-37.Cli-REST-C_Sharp/action/MarcarTareaFinalizadaAction.cs
using sdi3_37.Cli_REST_C_Sharp.dto;
using sdi3_37.Cli_REST_C_Sharp.util;
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

namespace sdi3_37.Cli_REST_C_Sharp.action
{
    class MarcarTareaFinalizadaAction : Action
    {
        private static String json;

        public MarcarTareaFinalizadaAction ()
        {

        }

        public void execute()
        {
            Console.WriteLine("------MARCAR TAREA COMO FINALIZADA------");

            //Pedir datos
            Console.Write("Id de la tarea que desea finalizar: ");
            var tareaIdEntrada = Console.ReadLine();

            long tareaId;
            if(long.TryParse(tareaIdEntrada, out tareaId))
            {
                buscarTarea(tareaId).Wait();
                List<Tarea> tarea = Util.parsearTareas(json);

                if(tarea[0] != null && tarea[0].userId == Sesion.Instance.User.id)
                {
                    if(tarea[0].finished == new DateTime(1970, 1, 1))
                    {
                        finalizarTarea(tarea[0]).Wait();
                    }
                    else
                    {
                        Console.WriteLine("\tLa tarea con id " + tareaId + " ya estaba finalizada");
                    }
                }
                else
                {
                    Console.WriteLine("\tNo existe la tarea con id " + tareaId);
                }
            }
            else
            {
                Console.WriteLine("\tId introducido inválido");
            }

            Console.WriteLine("----------------------------------------");
        }

        static async Task buscarTarea (long tareaId)
        {
            using (var cliente = new HttpClient())
            {
                cliente.BaseAddress = new Uri(Sesion.Instance.URL);
                cliente.DefaultRequestHeaders.Accept.Clear();
                cliente.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

                HttpResponseMessage response = await cliente.GetAsync("buscarTarea/id="+tareaId);
                if (response.IsSuccessStatusCode)
                {
                    json = await response.Content.ReadAsStringAsync();                    
                }
            }
        } 

        static async Task finalizarTarea (Tarea tarea)
        {
            using (var cliente = new HttpClient())
            {
                cliente.BaseAddress = new Uri(Sesion.Instance.URL);
                cliente.DefaultRequestHeaders.Accept.Clear();
                cliente.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

                var json = Util.parsearJson(tarea); 
                var stringContent = new StringContent(json, Encoding.UTF8, "application/json");
                HttpResponseMessage response = await cliente.PutAsync("finalizarTarea", stringContent);
                if (response.IsSuccessStatusCode)
                {
                    Console.WriteLine("\tTarea con id " + tarea.id + " marcada como finalizada correctamente");
                }
                else
                {
                    Console.WriteLine("\tHubo algún problema al marcar como finalizada la tarea con id " + tarea.id);
                }
            }
        }
    }
}
